### PR TITLE
mount shared mounts last

### PR DESCRIPTION
**Pull request by icewind1991**
That way the file name conflict resolution can work correctly

### CODE REVIEW

**Review by LukasReschke — COMMENTED** [review]
Looks good to me. I'll look into integrationstests here. Stay tuned :wink:

**Comment by LukasReschke** [comment]
@icewind1991 Tests seem to fail as per https://drone.nextcloud.com/nextcloud/server/5153/49

**Comment by icewind1991** [comment]
Fixed the tests

File: lib/private/Files/Config/MountProviderCollection.php [hunk 1]
```diff
@@ -84,6 +85,37 @@ public function getMountsForUser(IUser $user) {
 		}, array());
 	}
 
+	public function addMountForUser(IUser $user, IMountManager $mountManager) {
+		// shared mount provider gets to go last since it needs to know existing files
+		// to check for name collisions
+		$firstMounts = [];
+		$firstProviders = array_filter($this->providers, function (IMountProvider $provider) {
+			return (get_class($provider) !== 'OCA\Files_Sharing\MountProvider');
```
Comment (rullzer) [inline lib/private/Files/Config/MountProviderCollection.php]: why not use the `::class` ?

Comment (icewind1991) [inline lib/private/Files/Config/MountProviderCollection.php]: that would trigger autoloading, which fails if the app is disabled

Comment (icewind1991) [inline lib/private/Files/Config/MountProviderCollection.php]: which is also why `get_class` is used in the first place and not `instanceof`

Comment (rullzer) [inline lib/private/Files/Config/MountProviderCollection.php]: aaah ok fair enough. Lets do this then!

**Review by rullzer — APPROVED** [review]

**Comment by MorrisJobke** [comment]
> Looks good to me. I'll look into integrationstests here. Stay tuned 😉

What is the status of this?

File: lib/private/Files/Config/MountProviderCollection.php (hunk 1 above)
Comment (ChristophWurst) [inline lib/private/Files/Config/MountProviderCollection.php]: > that would trigger autoloading, which fails if the app is disabled

IIRC it wouldn't. I remember discussing this with @nickvergessen one time and we found out it's just syntactical sugar.

Comment (nickvergessen) [inline lib/private/Files/Config/MountProviderCollection.php]: Calling ::class will of course load the class ;)

**Comment by icewind1991** [comment]
Added integration test

File: lib/private/Files/Config/MountProviderCollection.php (hunk 1 above)
Comment (ChristophWurst) [inline lib/private/Files/Config/MountProviderCollection.php]: To verify, I've tested the following code:
```php
<?php

namespace X;

use A\B;

echo B::class;
```
which prints `A\B`. No exception thrown, hence I can assume that autoloading was not triggered.

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍

**Pull request merged**